### PR TITLE
A Nvidia gpu device plugin implementation for cos.

### PR DESCRIPTION
**Pull request by jiayingz**

### CODE REVIEW

File: device_plugins/nvidia_gpu/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+FROM ubuntu:16.04
+
+# Disable prompts from apt
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN apt-get -qq update
+RUN apt-get install -qq pciutils gcc g++ git make dpkg-dev bc module-init-tools curl
+
+RUN mkdir /lakitu-kernel
+RUN git clone https://chromium.googlesource.com/chromiumos/third_party/kernel /lakitu-kernel
```
Comment (vishh) [inline device_plugins/nvidia_gpu/Dockerfile]: Why is this needed?

File: device_plugins/nvidia_gpu/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM ubuntu:16.04
+
+# Disable prompts from apt
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN apt-get -qq update
+RUN apt-get install -qq pciutils gcc g++ git make dpkg-dev bc module-init-tools curl
```
Comment (vishh) [inline device_plugins/nvidia_gpu/Dockerfile]: Are all these tools necessary?

File: device_plugins/nvidia_gpu/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+RUN mkdir /lakitu-kernel
+RUN git clone https://chromium.googlesource.com/chromiumos/third_party/kernel /lakitu-kernel
+
+COPY installer.sh /usr/bin/nvidia-installer.sh
+RUN chmod a+x /usr/bin/nvidia-installer.sh
+COPY nvidia_gpu /usr/bin/device_plugins
+RUN chmod a+x /usr/bin/device_plugins
```
Comment (vishh) [inline device_plugins/nvidia_gpu/Dockerfile]: ideally, the device plugin itself is a static binary which makes the container small.

File: device_plugins/nvidia_gpu/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+# Disable prompts from apt
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN apt-get -qq update
+RUN apt-get install -qq pciutils gcc g++ git make dpkg-dev bc module-init-tools curl
+
+RUN mkdir /lakitu-kernel
+RUN git clone https://chromium.googlesource.com/chromiumos/third_party/kernel /lakitu-kernel
+
+COPY installer.sh /usr/bin/nvidia-installer.sh
```
Comment (vishh) [inline device_plugins/nvidia_gpu/Dockerfile]: Let's separate the installation from device plugins.

File: device_plugins/nvidia_gpu/device-plugin-daemonset.yaml [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      - name: nvidia-overlay
+        hostPath:
+          path: /home/kubernetes/bin/nvidia
+      - name: os-release
+        hostPath:
+          path: /etc/os-release
+      - name: sysrq
+        hostPath:
+          path: /proc/sysrq-trigger
+      containers:
```
Comment (vishh) [inline device_plugins/nvidia_gpu/device-plugin-daemonset.yaml]: Let's add an init container to installer the drivers.

Comment (vishh) [reply device_plugins/nvidia_gpu/device-plugin-daemonset.yaml]: this implicitly requires installer to be in a different container.

File: device_plugins/nvidia_gpu/installer.sh [hunk 6]
```diff
@@ -0,0 +1,230 @@
+#!/bin/bash
+
```
Comment (vishh) [inline device_plugins/nvidia_gpu/installer.sh]: Leave installer out of plugin.

File: device_plugins/nvidia_gpu/nvidia_gpu.go [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+	files, err := ioutil.ReadDir(devDirectory)
+	if err != nil {
+		return err
+	}
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
+		if reg.MatchString(f.Name()) {
+			fmt.Printf("Found Nvidia GPU %q\n", f.Name())
```
Comment (vishh) [inline device_plugins/nvidia_gpu/nvidia_gpu.go]: nit: use glog.

File: device_plugins/nvidia_gpu/nvidia_gpu.go [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+	return pluginapi.Healthy
+}
+
+// Discovers Nvidia GPU devices, installs device drivers, and sets up device
+// access environment.
+func (ngm *nvidiaGPUManager) Start() error {
+	// Install Nvidia device drivers.
+	// TODO: turn this into a go library to get better error reporting.
+	fmt.Printf("Run installer script\n")
+	cmd := exec.Command("/bin/sh", "-c", "/usr/bin/nvidia-installer.sh")
```
Comment (vishh) [inline device_plugins/nvidia_gpu/nvidia_gpu.go]: Installation is done out of band.

File: device_plugins/nvidia_gpu/nvidia_gpu.go [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+		}
+		if changed {
+			resp := new(pluginapi.ListAndWatchResponse)
+			for _, dev := range ngm.devices {
+				resp.Devices = append(resp.Devices, &pluginapi.Device{dev.ID, dev.Health})
+			}
+			fmt.Printf("ListAndWatch: send devices %v", resp)
+			if err := stream.Send(resp); err != nil {
+				fmt.Printf("device-plugin: cannot update device states: %v\n", err)
+				ngm.grpcServer.Stop()
```
Comment (vishh) [inline device_plugins/nvidia_gpu/nvidia_gpu.go]: What happens if the server is stopped?

File: device_plugins/nvidia_gpu/nvidia_gpu.go [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+		for _, d := range ngm.defaultDevices {
+			devRuntime.Devices = append(devRuntime.Devices, &pluginapi.DeviceSpec{
+				HostPath:      d,
+				ContainerPath: d,
+				Permissions:   "mrw",
+			})
+		}
+		devRuntime.Mounts = append(devRuntime.Mounts, &pluginapi.Mount{
+			ContainerPath: "/usr/local/nvidia/lib64",
+			HostPath:  "/home/kubernetes/bin/nvidia/lib",
```
Comment (vishh) [inline device_plugins/nvidia_gpu/nvidia_gpu.go]: Can we use const to represent base paths? Make the base paths configurable to make the plugin work outside of COS.

File: device_plugins/nvidia_gpu/nvidia_gpu.go (hunk 9 above)
Comment (jiayingz) [reply device_plugins/nvidia_gpu/nvidia_gpu.go]: device plugin will re-register with kubelet.

File: device_plugins/nvidia_gpu/nvidia_gpu.go (hunk 10 above)
Comment (jiayingz) [reply device_plugins/nvidia_gpu/nvidia_gpu.go]: done.

File: device_plugins/nvidia_gpu/nvidia_gpu.go (hunk 8 above)
Comment (jiayingz) [reply device_plugins/nvidia_gpu/nvidia_gpu.go]: Added this in todo.

File: device_plugins/nvidia_gpu/installer.sh (hunk 6 above)
Comment (jiayingz) [reply device_plugins/nvidia_gpu/installer.sh]: removed.

File: device_plugins/nvidia_gpu/device-plugin-daemonset.yaml (hunk 5 above)
Comment (jiayingz) [reply device_plugins/nvidia_gpu/device-plugin-daemonset.yaml]: as we discussed offline, we will do this separately in a later PR.

File: device_plugins/nvidia_gpu/nvidia_gpu.go (hunk 7 above)
Comment (jiayingz) [reply device_plugins/nvidia_gpu/nvidia_gpu.go]: I used fm because it is easier to see the output from kubectl logs. Is there a way to get glog output from kubectl logs?

File: device_plugins/nvidia_gpu/Makefile [hunk 11]
```diff
@@ -0,0 +1,25 @@
+# Copyright 2017 The Kubernetes Authors.
```
Comment (vishh) [inline device_plugins/nvidia_gpu/Makefile]: Can you move this Makefile to the top of the repository?

File: device_plugins/nvidia_gpu/device-plugin-daemonset.yaml [hunk 12]
```diff
@@ -0,0 +1,55 @@
+apiVersion: extensions/v1beta1
```
Comment (vishh) [inline device_plugins/nvidia_gpu/device-plugin-daemonset.yaml]: Needs a license banner I think

File: device_plugins/nvidia_gpu/device-plugin-daemonset.yaml [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+kind: DaemonSet
```
Comment (vishh) [inline device_plugins/nvidia_gpu/device-plugin-daemonset.yaml]: Add a comment describing how this daemonset is expected to be used and what it will do.

File: device_plugins/nvidia_gpu/nvidia_gpu.go [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+	// Optional device.
+	nvidiaUVMToolsDevice string = "/dev/nvidia-uvm-tools"
+	devDirectory                = "/dev"
+	nvidiaDeviceRE              = `^nvidia[0-9]*$`
+
+	// Device plugin settings.
+	devicePluginMountPath = "/device-plugin"
+	kubeletEndpoint       = "kubelet.sock"
+	pluginEndpointPrefix  = "nvidiaGPU"
+	resourceName          = "nvidia.com/GPU"
```
Comment (vishh) [inline device_plugins/nvidia_gpu/nvidia_gpu.go]: the resource name is `gpu` and not `GPU`. It's more convenient to type lower case letters.

File: device_plugins/nvidia_gpu/nvidia_gpu.go [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+	ContainerPathPrefix   = "/usr/local/nvidia"
```
Comment (vishh) [inline device_plugins/nvidia_gpu/nvidia_gpu.go]: note as a next step, we need to also inject ldconfig - relevant Nvidia code [here](https://github.com/NVIDIA/nvidia-docker/blob/3422bca272d64600ca15874d96e4e511b669c7c4/src/nvidia/volumes.go#L234). ideally we should re-use their code.

File: nvidia_gpu/BUILD [hunk 16]
```diff
@@ -0,0 +1,57 @@
+# Copyright 2017 Google Inc. All rights reserved.
```
Comment (vishh) [inline nvidia_gpu/BUILD]: Do we need bazel build for this repo?

File: nvidia_gpu/device-plugin-daemonset.yaml [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+        hostPath:
+          path: /home/kubernetes/bin/nvidia
+      - name: os-release
+        hostPath:
+          path: /etc/os-release
+      - name: sysrq
+        hostPath:
+          path: /proc/sysrq-trigger
+      containers:
+      - image: "gcr.io/google_containers/device-plugin-gpu:v0.3"
```
Comment (vishh) [inline nvidia_gpu/device-plugin-daemonset.yaml]: nit: Use an image digest. Tags are risky.

File: nvidia_gpu/Makefile [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+TAG?=v0.3
+REGISTRY?=gcr.io/google_containers
+IMAGE=device-plugin-gpu
+build:
```
Comment (vishh) [inline nvidia_gpu/Makefile]: Move this to the top level Makefile

**Comment by vishh** [comment]
Feel free to self merge once you have fixed the tests
